src/main/java/org/wildfly/security/http/HttpAuthenticator.java

```diff
@@ -98,12 +110,20 @@ public static Builder builder() {
         private volatile HttpServerMechanismsResponder successResponder;
 
         private boolean authenticate() throws HttpAuthenticationException {
+            return evaluate((ExceptionConsumer<HttpServerAuthenticationMechanism, HttpAuthenticationException>) mechanism -> mechanism.evaluateRequest(this));
+        }
+
+        private boolean logout() throws HttpAuthenticationException {
+            return evaluate((ExceptionConsumer<HttpServerAuthenticationMechanism, HttpAuthenticationException>) mechanism -> mechanism.evaluateLogoutRequest(this));
+        }
+
+        private boolean evaluate(ExceptionConsumer<HttpServerAuthenticationMechanism, HttpAuthenticationException> mechanism) throws HttpAuthenticationException {
             List<HttpServerAuthenticationMechanism> authenticationMechanisms = mechanismSupplier.get();
             responders = new ArrayList<>(authenticationMechanisms.size());
             try {
                 for (HttpServerAuthenticationMechanism nextMechanism : authenticationMechanisms) {
                     currentMechanism = nextMechanism;
-                    nextMechanism.evaluateRequest(this);
+                    mechanism.accept(nextMechanism);
 
                     if (isAuthenticated()) {
```

Comment:
It depends, this is about managing logout notifications usually sent as a result of a programmatic logout from application/container.

Mechanisms are free to act however they want at during logout. They can re-authenticate the user and invalidate any mech specific state associated with the identity (so subsequent requests will always be unauthenticated), they can only check internal state and send a response to the client if they want to force a flow, etc.It depends, this is about managing logout notifications usually sent as a result of a programmatic logout from application/container.

Mechanisms are free to act however they want at during logout. They can re-authenticate the user and invalidate any mech specific state associated with the identity (so subsequent requests will always be unauthenticated), they can only check internal state and send a response to the client if they want to force a flow, etc.